cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java

```diff
@@ -197,6 +208,76 @@ private void upgradeSystemDatasets(Injector injector) throws Exception {
+    Version version = findMetricsTableVersion(injector.getInstance(CConfiguration.class),
+                                                  injector.getInstance(Configuration.class), metricHBaseTableUtil);
+    if (version != null) {
+      DatasetFramework framework = createRegisteredDatasetFramework(injector);
+      // migrate metrics data
+      DefaultMetricDatasetFactory.migrateData(injector.getInstance(CConfiguration.class), framework, version);
+    }
+  }
+
+  @Nullable
+  private Version findMetricsTableVersion(CConfiguration cConf, Configuration hConf,
+                                                      MetricHBaseTableUtil metricHBaseTableUtil) {
+
+
+    // Figure out what is the latest working version of CDAP
+    // 1) if latest is 2.8.x - nothing to do, if "pre-2.8", proceed to next step.
+    // 2) find a most recent metrics table: start by looking for 2.7, then 2.6
+    // 3) if we find 2.7 - we will migrate data from 2.7 table, if not - migrate data from 2.6 metrics table
+    // todo - use UpgradeTool to figure out if version is 2.8.x, return if it is 2.8.x
+
+    String tableName27 = cConf.get(MetricsConstants.ConfigKeys.METRICS_TABLE_PREFIX,
+                                   UpgradeMetricsConstants.DEFAULT_METRICS_TABLE_PREFIX) + ".agg";
+
+    // versions older than 2.7, has two metrics table, identified by system and user prefix
+    String tableName26 = "system." + tableName27;
+    DefaultDatasetNamespace defaultDatasetNamespace = new DefaultDatasetNamespace(cConf);
+    String metricsEntityTable26 = defaultDatasetNamespace.namespace(new Id.Namespace(Constants.SYSTEM_NAMESPACE),
+                                                                    tableName26);
+    String metricsEntityTable27 = defaultDatasetNamespace.namespace(new Id.Namespace(Constants.SYSTEM_NAMESPACE),
+                                                                    tableName27);
+    HBaseAdmin hAdmin = null;
+    Version version = null;
+    try {
+      hAdmin = new HBaseAdmin(hConf);
+      for (HTableDescriptor desc : hAdmin.listTables()) {
+        if (desc.getNameAsString().equals(metricsEntityTable27)) {
+          version = metricHBaseTableUtil.getVersion(desc);
+          version = verifyVersion(Version.VERSION_2_7, version);
+          if (version == null) {
+            return null;
+          }
+          break;
+        }
+        if (desc.getNameAsString().equals(metricsEntityTable26)) {
+          version = metricHBaseTableUtil.getVersion(desc);
+          version = verifyVersion(Version.VERSION_2_6_OR_LOWER, version);
+          if (version == null) {
+            return null;
+          }
+        }
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+    return version;
+  }
+
+  private Version verifyVersion(Version expected, Version actual) {
+    if (expected != actual) {
+      System.out.println("Unable to migrate, Version mismatch, Expected Version " + expected +
```

Comment:
shouldn't you say: "version detected based on table name does not match table configuration"?